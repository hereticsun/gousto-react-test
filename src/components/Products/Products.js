import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Product from '../Product';
import SearchBar from '../SearchBar';
import { fetchProducts } from '../../actions/products';

export class Products extends Component {
  componentDidMount() {
    this.props.fetchProducts();
  }

  productList(){
    if(Boolean(this.props.selectedCategory.id)) {
      const filteredProducts = this.props.products.filter(
        product => product.categories.some(category => category.id === this.props.selectedCategory.id)
      );

      if(Boolean(this.props.searchTerm)) {
        const searchedProducts = filteredProducts.filter(
          product => {
            const lowerTitle = product.title.toLowerCase();
            const lowerDesc = product.description.toLowerCase();
            const q = this.props.searchTerm.toLowerCase();
            const test = lowerTitle.includes(q) || lowerDesc.includes(q);
            return test;
          }
        );
        return searchedProducts;
      }
      return filteredProducts;
    }
    if(Boolean(this.props.searchTerm)) {
      const searchedProducts = this.props.products.filter(
        product => {
          const lowerTitle = product.title.toLowerCase();
          const lowerDesc = product.description.toLowerCase();
          const q = this.props.searchTerm.toLowerCase();
          const test = lowerTitle.includes(q) || lowerDesc.includes(q);
          return test;
        }
      );
      return searchedProducts;
    }
    return this.props.products;
  }

  render() {
    const productHeading = Boolean(this.props.selectedCategory.title) ?
      `Products in ${this.props.selectedCategory.title}` :
      'Products';
    return (
      <section className="products">
        <h2 className="products__heading">{productHeading}</h2>
        <SearchBar />
        {this.props.products.length > 0 && (
          <ul className="products__list">
            {this.productList().map(product => (
              <Product key={product.id} product={product} />
            ))}
          </ul>
        )}
      </section>
    );
  }
}

function mapStateToProps(state) {
  return {
      products: state.products,
      selectedCategory: state.selectedCategory,
      searchTerm: state.searchTerm,
  };
}

Products.propTypes = {
  selectedCategory: PropTypes.object,
  products: PropTypes.array,
  searchTerm: PropTypes.string,
};

export default connect(mapStateToProps, { fetchProducts })(Products);
