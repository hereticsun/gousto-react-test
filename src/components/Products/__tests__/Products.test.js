import React from 'react';
import { shallow } from 'enzyme';

import {Products} from '../Products';

describe(`<Products />`, () => {
  describe('Given that I am a user', () => {
    describe('when I land on the main page', () => {
      const props = {
        products: [
          {
            id: '1',
            title: 'Borsao Macabeo',
            categories: [{id: 1}],
          },
          {
            id: '2',
            title: 'Love Shortie All Butter Shortbread',
            categories: [],
          },
        ],
        fetchProducts: jest.fn()
      };
      const element = shallow(<Products {...props} />);

      it('displays the list of product titles', () => {
        expect(element).toMatchSnapshot();
        expect(element.find('Product')).toHaveLength(props.products.length);
      });
    });
  });
});
