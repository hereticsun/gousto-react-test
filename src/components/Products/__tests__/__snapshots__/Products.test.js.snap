// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> Given that I am a user when I land on the main page displays the list of product titles 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Products
    fetchProducts={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    products={
      Array [
        Object {
          "categories": Array [
            Object {
              "id": 1,
            },
          ],
          "id": "1",
          "title": "Borsao Macabeo",
        },
        Object {
          "categories": Array [],
          "id": "2",
          "title": "Love Shortie All Butter Shortbread",
        },
      ]
    }
    selectedCategory={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="products__heading"
        >
          Products
        </h2>,
        <Connect(SearchBar) />,
        <ul
          className="products__list"
        >
          <Product
            product={
              Object {
                "categories": Array [
                  Object {
                    "id": 1,
                  },
                ],
                "id": "1",
                "title": "Borsao Macabeo",
              }
            }
          />
          <Product
            product={
              Object {
                "categories": Array [],
                "id": "2",
                "title": "Love Shortie All Butter Shortbread",
              }
            }
          />
        </ul>,
      ],
      "className": "products",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Products",
          "className": "products__heading",
        },
        "ref": null,
        "rendered": "Products",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Product
              product={
                Object {
                  "categories": Array [
                    Object {
                      "id": 1,
                    },
                  ],
                  "id": "1",
                  "title": "Borsao Macabeo",
                }
              }
            />,
            <Product
              product={
                Object {
                  "categories": Array [],
                  "id": "2",
                  "title": "Love Shortie All Butter Shortbread",
                }
              }
            />,
          ],
          "className": "products__list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "product": Object {
                "categories": Array [
                  Object {
                    "id": 1,
                  },
                ],
                "id": "1",
                "title": "Borsao Macabeo",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "product": Object {
                "categories": Array [],
                "id": "2",
                "title": "Love Shortie All Butter Shortbread",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="products__heading"
          >
            Products
          </h2>,
          <Connect(SearchBar) />,
          <ul
            className="products__list"
          >
            <Product
              product={
                Object {
                  "categories": Array [
                    Object {
                      "id": 1,
                    },
                  ],
                  "id": "1",
                  "title": "Borsao Macabeo",
                }
              }
            />
            <Product
              product={
                Object {
                  "categories": Array [],
                  "id": "2",
                  "title": "Love Shortie All Butter Shortbread",
                }
              }
            />
          </ul>,
        ],
        "className": "products",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Products",
            "className": "products__heading",
          },
          "ref": null,
          "rendered": "Products",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Product
                product={
                  Object {
                    "categories": Array [
                      Object {
                        "id": 1,
                      },
                    ],
                    "id": "1",
                    "title": "Borsao Macabeo",
                  }
                }
              />,
              <Product
                product={
                  Object {
                    "categories": Array [],
                    "id": "2",
                    "title": "Love Shortie All Butter Shortbread",
                  }
                }
              />,
            ],
            "className": "products__list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "product": Object {
                  "categories": Array [
                    Object {
                      "id": 1,
                    },
                  ],
                  "id": "1",
                  "title": "Borsao Macabeo",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "product": Object {
                  "categories": Array [],
                  "id": "2",
                  "title": "Love Shortie All Butter Shortbread",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
